In [1]:
import pandas as pd

In [2]:
dri = pd.read_csv("//Drink_Mixer.csv")

In [3]:
dri.head()

Name  \
0        '61 Imperial recipe   
1         .50 Caliber recipe   
2                 007 recipe   
3             10 Deep recipe   
4  100 Miles per Hour recipe   

                                         Ingredients  \
0  ['1/2 oz white rum', '1/2 oz vodka', '1/2 oz 1...   
1  ["2 oz Jack Daniel's® Tennessee whiskey", '2 o...   
2  ['3 oz Stoli® Ohranj vodka', '1 1/2 ozfreshly ...   
3  ['1.5 oz vodka', '1 oz Malibu® coconut rum', '...   
4  ['1/2 oz Wild Turkey® bourbon whiskey', '1/2 o...   

                                              Making  \
0  In a tall chimney glass or hurricane glass, fi...   
1  Pour the Jack Daniel's Tennessee whiskey, Jim ...   
2                  Serve on ice in a highball glass.   
3  All right start with 1/2 cup of ice mix the vo...   
4  Mix all ingredients together over ice in a hig...   

                                           Link            Glass Type Base  
0     http://www.drinksmixer.com/drink8428.html  Hurricane Glass  NaN  NaN  
1     http://www.drinksmixer.com/drink9993.html   Highball Glass  NaN  NaN  
2    http://www.drinksmixer.com/drink11090.html   Highball Glass  NaN  NaN  
3  http://www.drinksmixer.com/drinkuv17048.html              Cup  NaN  NaN  
4    http://www.drinksmixer.com/drink10390.html   Highball Glass  NaN  NaN

In [4]:
dri.describe()

Name  \
count                 16990   
unique                16649   
top     Fire and Ice recipe   
freq                      3   

                                              Ingredients  \
count                                               16990   
unique                                              16725   
top     ["1/2 oz Jack Daniel's® Tennessee whiskey", '1...   
freq                                                    4   

                                                   Making  \
count                                               16990   
unique                                              15202   
top     Shake ingredients in a cocktail shaker with ic...   
freq                                                  142   

                                                Link           Glass     Type  \
count                                          16990           16626     4004   
unique                                         16818              38        3   
top     http://www.drinksmixer.com/drinkl13m214.html  Cocktail Glass  Shooter   
freq                                               2            3409     2672   

            Base  
count        652  
unique         2  
top     Beer&Ale  
freq         342

In [682]:
dricopy = dri.copy()

칵테일 이름 뒤에 붙은 " recipe" 없애주기

In [683]:
for i in range(len(dricopy)):
    if "recipe" not in dricopy["Name"][i]:
        print(i)

In [684]:
for i in range(len(dricopy)):
    if dricopy["Name"][i][-7:]!=" recipe":
        print(i)

In [685]:
for i in range(len(dricopy)):
    if dricopy["Name"][i][-7:]==" recipe":
        dricopy["Name"][i] = dricopy["Name"][i][:-7]

In [686]:
# 확인
dricopy.head()

Name                                        Ingredients  \
0        '61 Imperial  ['1/2 oz white rum', '1/2 oz vodka', '1/2 oz 1...   
1         .50 Caliber  ["2 oz Jack Daniel's® Tennessee whiskey", '2 o...   
2                 007  ['3 oz Stoli® Ohranj vodka', '1 1/2 ozfreshly ...   
3             10 Deep  ['1.5 oz vodka', '1 oz Malibu® coconut rum', '...   
4  100 Miles per Hour  ['1/2 oz Wild Turkey® bourbon whiskey', '1/2 o...   

                                              Making  \
0  In a tall chimney glass or hurricane glass, fi...   
1  Pour the Jack Daniel's Tennessee whiskey, Jim ...   
2                  Serve on ice in a highball glass.   
3  All right start with 1/2 cup of ice mix the vo...   
4  Mix all ingredients together over ice in a hig...   

                                           Link            Glass Type Base  
0     http://www.drinksmixer.com/drink8428.html  Hurricane Glass  NaN  NaN  
1     http://www.drinksmixer.com/drink9993.html   Highball Glass  NaN  NaN  
2    http://www.drinksmixer.com/drink11090.html   Highball Glass  NaN  NaN  
3  http://www.drinksmixer.com/drinkuv17048.html              Cup  NaN  NaN  
4    http://www.drinksmixer.com/drink10390.html   Highball Glass  NaN  NaN

### 중복 확인 및 제거

전체 중복 확인

In [687]:
dri_ent_dup = dricopy.duplicated(subset=None, keep=False)
len(dri_ent_dup[dri_ent_dup==True])

4

이름+재료만 중복 확인

In [688]:
dri_2_dup = dricopy.duplicated(subset=["Name", "Ingredients"], keep=False)
len(dri_2_dup[dri_2_dup==True])

384

링크만 중복 확인

In [689]:
dri_link_dup = dricopy.duplicated(subset=["Link"], keep=False)
len(dri_link_dup[dri_link_dup==True])

344

1. 일단 전체중복은 유니크만 남기고 drop하기

In [690]:
dri_ent_dup_ind = []
for i in range(len(dricopy)):
    if dri_ent_dup[i]==True:
        dri_ent_dup_ind.append(i)
print(dri_ent_dup_ind)

[12418, 12428, 14623, 14633]


In [691]:
for i in dri_ent_dup_ind:
    print(dricopy["Name"][i])

Alien Nipple #2
Alien Nipple #2
Shocker
Shocker


In [692]:
dricopy.drop_duplicates(subset=None, keep='first', inplace=True, ignore_index=True)

In [693]:
dri_ent_dup = dricopy.duplicated(subset=None, keep=False)
len(dri_ent_dup[dri_ent_dup==True])

0

In [694]:
dricopy

Name                                        Ingredients  \
0            '61 Imperial  ['1/2 oz white rum', '1/2 oz vodka', '1/2 oz 1...   
1             .50 Caliber  ["2 oz Jack Daniel's® Tennessee whiskey", '2 o...   
2                     007  ['3 oz Stoli® Ohranj vodka', '1 1/2 ozfreshly ...   
3                 10 Deep  ['1.5 oz vodka', '1 oz Malibu® coconut rum', '...   
4      100 Miles per Hour  ['1/2 oz Wild Turkey® bourbon whiskey', '1/2 o...   
...                   ...                                                ...   
16983          Voodoo Tea  ['1 shot white rum', '1/2 shot Blue Curacao li...   
16984     Whiskey Punjabi  ['2 - 3 ozpremium whiskey', '8 - 12 ozunsweete...   
16985         Wild Beaver  ["3 oz Jack Daniel's® Tennessee whiskey", '3 o...   
16986     Xanthor's Pants  ['2 oz Jameson® Irish whiskey', "2 oz Bailey's...   
16987               Zorro  ['2 cl sambuca', "2 cl Bailey's® Irish cream",...   

                                                  Making  \
0      In a tall chimney glass or hurricane glass, fi...   
1      Pour the Jack Daniel's Tennessee whiskey, Jim ...   
2                      Serve on ice in a highball glass.   
3      All right start with 1/2 cup of ice mix the vo...   
4      Mix all ingredients together over ice in a hig...   
...                                                  ...   
16983  Pour rum and curacao into a highball glass. Fi...   
16984                                         Mix, stir.   
16985  Combine all ingredients in equal parts in a co...   
16986  Mix 1 part Scotch with 1 part Bailey's (approx...   
16987  Pour all ingredients into an irish coffee cup....   

                                               Link                Glass Type  \
0         http://www.drinksmixer.com/drink8428.html      Hurricane Glass  NaN   
1         http://www.drinksmixer.com/drink9993.html       Highball Glass  NaN   
2        http://www.drinksmixer.com/drink11090.html       Highball Glass  NaN   
3      http://www.drinksmixer.com/drinkuv17048.html                  Cup  NaN   
4        http://www.drinksmixer.com/drink10390.html       Highball Glass  NaN   
...                                             ...                  ...  ...   
16983     http://www.drinksmixer.com/drink6586.html       Highball Glass  NaN   
16984  http://www.drinksmixer.com/drinkvw15078.html  Old-Fashioned Glass  NaN   
16985     http://www.drinksmixer.com/drink8113.html         Beer Pilsner  NaN   
16986  http://www.drinksmixer.com/drinkrq15073.html           Coffee Mug  NaN   
16987     http://www.drinksmixer.com/drink6609.html     Irish Coffee Cup  NaN   

             Base  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
...           ...  
16983  Coffee&Tee  
16984  Coffee&Tee  
16985  Coffee&Tee  
16986  Coffee&Tee  
16987  Coffee&Tee  

[16988 rows x 7 columns]

2. 링크 중복 레시피들의 유니크 링크 별로 glass, type, base를 합친 후 하나씩만 남기고 drop하기

In [695]:
dri_link_dup = dricopy.duplicated(subset=["Link"], keep=False)
len(dri_link_dup[dri_link_dup==True])

340

In [696]:
dri_link_dup_ind = []
for i in range(len(dricopy)):
    if dri_link_dup[i]==True:
        dri_link_dup_ind.append(i)
print(len(dri_link_dup_ind))

340


In [697]:
link_unique = []
for i in dri_link_dup_ind:
    if dricopy["Link"][i] not in link_unique:
        link_unique.append(dricopy["Link"][i])
print(len(link_unique))

170


In [698]:
link_other_collection = {}
for i in range(len(dricopy)):
    if dricopy["Link"][i] in link_unique:
        if dricopy["Link"][i] not in list(link_other_collection.keys()):
            link_other_collection[dricopy["Link"][i]] = {}
            if str(dricopy["Glass"][i])!='nan':
                link_other_collection[dricopy["Link"][i]]["Glass"] = [dricopy["Glass"][i]]
            else:
                link_other_collection[dricopy["Link"][i]]["Glass"] = []
            if str(dricopy["Type"][i])!='nan':
                link_other_collection[dricopy["Link"][i]]["Type"] = [dricopy["Type"][i]]
            else:
                link_other_collection[dricopy["Link"][i]]["Type"] = []
            if str(dricopy["Base"][i])!='nan':
                link_other_collection[dricopy["Link"][i]]["Base"] = [dricopy["Base"][i]]
            else:
                link_other_collection[dricopy["Link"][i]]["Base"] = []
        else:
            if dricopy["Glass"][i] not in link_other_collection[dricopy["Link"][i]]["Glass"] and str(dricopy["Glass"][i])!='nan':
                link_other_collection[dricopy["Link"][i]]["Glass"].append(dricopy["Glass"][i])
            if dricopy["Type"][i] not in link_other_collection[dricopy["Link"][i]]["Type"] and str(dricopy["Type"][i])!='nan':
                link_other_collection[dricopy["Link"][i]]["Type"].append(dricopy["Type"][i])
            if dricopy["Base"][i] not in link_other_collection[dricopy["Link"][i]]["Base"] and str(dricopy["Base"][i])!='nan':
                link_other_collection[dricopy["Link"][i]]["Base"].append(dricopy["Base"][i])
print(link_other_collection)

{'http://www.drinksmixer.com/drink3402.html': {'Glass': ['Shot Glass'], 'Type': ['Shooter'], 'Base': []}, 'http://www.drinksmixer.com/drink792.html': {'Glass': ['Shot Glass'], 'Type': ['Shooter'], 'Base': []}, 'http://www.drinksmixer.com/drink12u5t74.html': {'Glass': ['Irish Coffee Cup'], 'Type': [], 'Base': ['Coffee&Tee']}, 'http://www.drinksmixer.com/drink8452.html': {'Glass': ['Punch Bowl'], 'Type': ['Punch'], 'Base': []}, 'http://www.drinksmixer.com/drink11169.html': {'Glass': ['Shot Glass'], 'Type': ['Shooter'], 'Base': []}, 'http://www.drinksmixer.com/drink4140.html': {'Glass': ['Shot Glass'], 'Type': ['Shooter'], 'Base': []}, 'http://www.drinksmixer.com/drink7538.html': {'Glass': ['Beer Mug'], 'Type': [], 'Base': ['Beer&Ale']}, 'http://www.drinksmixer.com/drinkb1c2698.html': {'Glass': ['Collins Glass'], 'Type': ['Nonalcoholic'], 'Base': []}, 'http://www.drinksmixer.com/drinkvw12765.html': {'Glass': ['Cocktail Glass'], 'Type': [], 'Base': ['Coffee&Tee']}, 'http://www.drinksmixer.

In [699]:
len(link_other_collection)

170

In [700]:
link_other_collection_key = list(link_other_collection.keys())
for i in range(len(link_other_collection_key)):
    gathered_glasses = link_other_collection[link_other_collection_key[i]]["Glass"]
    if len(gathered_glasses)>=2:
        count = 0
        glass_glass_ind = []
        for j in len(gathered_glasses):
            if 'Glass' in gathered_glasses[j]:
                count += 1
                glass_glass_ind.append(j)
        if count>=2:
            replace_count = 0
            while replace_count<glass_glass_ind[-1]:
                gathered_glasses[replace_count] = gathered_glasses[replace_count].replace(' Glass', "")
                replace_count += 1
for j in dri_link_dup_ind:
    dricopy["Glass"][j] = " or ".join(link_other_collection[dricopy["Link"][j]]["Glass"])
    dricopy["Type"][j] = ", ".join(link_other_collection[dricopy["Link"][j]]["Type"])
    dricopy["Base"][j] = ", ".join(link_other_collection[dricopy["Link"][j]]["Base"])

In [701]:
dricopy.drop_duplicates(subset=["Link"], keep='first', inplace=True, ignore_index=True)

In [702]:
dri_link_dup = dricopy.duplicated(subset=["Link"], keep=False)
len(dri_link_dup[dri_link_dup==True])

0

In [703]:
dricopy

Name                                        Ingredients  \
0            '61 Imperial  ['1/2 oz white rum', '1/2 oz vodka', '1/2 oz 1...   
1             .50 Caliber  ["2 oz Jack Daniel's® Tennessee whiskey", '2 o...   
2                     007  ['3 oz Stoli® Ohranj vodka', '1 1/2 ozfreshly ...   
3                 10 Deep  ['1.5 oz vodka', '1 oz Malibu® coconut rum', '...   
4      100 Miles per Hour  ['1/2 oz Wild Turkey® bourbon whiskey', '1/2 o...   
...                   ...                                                ...   
16813          Voodoo Tea  ['1 shot white rum', '1/2 shot Blue Curacao li...   
16814     Whiskey Punjabi  ['2 - 3 ozpremium whiskey', '8 - 12 ozunsweete...   
16815         Wild Beaver  ["3 oz Jack Daniel's® Tennessee whiskey", '3 o...   
16816     Xanthor's Pants  ['2 oz Jameson® Irish whiskey', "2 oz Bailey's...   
16817               Zorro  ['2 cl sambuca', "2 cl Bailey's® Irish cream",...   

                                                  Making  \
0      In a tall chimney glass or hurricane glass, fi...   
1      Pour the Jack Daniel's Tennessee whiskey, Jim ...   
2                      Serve on ice in a highball glass.   
3      All right start with 1/2 cup of ice mix the vo...   
4      Mix all ingredients together over ice in a hig...   
...                                                  ...   
16813  Pour rum and curacao into a highball glass. Fi...   
16814                                         Mix, stir.   
16815  Combine all ingredients in equal parts in a co...   
16816  Mix 1 part Scotch with 1 part Bailey's (approx...   
16817  Pour all ingredients into an irish coffee cup....   

                                               Link                Glass Type  \
0         http://www.drinksmixer.com/drink8428.html      Hurricane Glass  NaN   
1         http://www.drinksmixer.com/drink9993.html       Highball Glass  NaN   
2        http://www.drinksmixer.com/drink11090.html       Highball Glass  NaN   
3      http://www.drinksmixer.com/drinkuv17048.html                  Cup  NaN   
4        http://www.drinksmixer.com/drink10390.html       Highball Glass  NaN   
...                                             ...                  ...  ...   
16813     http://www.drinksmixer.com/drink6586.html       Highball Glass  NaN   
16814  http://www.drinksmixer.com/drinkvw15078.html  Old-Fashioned Glass  NaN   
16815     http://www.drinksmixer.com/drink8113.html         Beer Pilsner  NaN   
16816  http://www.drinksmixer.com/drinkrq15073.html           Coffee Mug  NaN   
16817     http://www.drinksmixer.com/drink6609.html     Irish Coffee Cup  NaN   

             Base  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
...           ...  
16813  Coffee&Tee  
16814  Coffee&Tee  
16815  Coffee&Tee  
16816  Coffee&Tee  
16817  Coffee&Tee  

[16818 rows x 7 columns]

3. 이름+재료 중복 레시피들의 유니크 이름+재료 별로 Making, Glass, Base를 출력. 이걸 보고 바리에이션 넘버 부여 혹은 drop하기

In [704]:
dri_2_dup = dricopy.duplicated(subset=["Name", "Ingredients"], keep=False)
len(dri_2_dup[dri_2_dup==True])

40

In [705]:
dri_2_dup_ind = []
for i in range(len(dricopy)):
    if dri_2_dup[i]==True:
        dri_2_dup_ind.append(i)
print(len(dri_2_dup_ind))

40


In [706]:
name_ing_unique = []
name_ing_ind_col = {}
for i in dri_2_dup_ind:
    if dricopy["Name"][i]+dricopy["Ingredients"][i] not in name_ing_unique:
        name_ing_unique.append(dricopy["Name"][i]+dricopy["Ingredients"][i])
        name_ing_ind_col[dricopy["Name"][i]+dricopy["Ingredients"][i]] = [i]
    else:
        name_ing_ind_col[dricopy["Name"][i]+dricopy["Ingredients"][i]].append(i)
print(name_ing_ind_col)

{"Caribbean Martini['1 1/2 oz Stoli® Vanil vodka', '3/4 oz Malibu® coconut rum', '1 splash pineapple juice']": [2527, 2528], "Christmas in a Cup['1 oz spiced rum', '1 oz Aftershock® Hot & Cool cinnamon schnapps', '2 oz Dr. Pepper® soda']": [2972, 2973], "Crimson Sunset['2 oz gin', '2 tsp lemon juice', '1/2 tsp grenadine syrup', '1/2 oz tawny port']": [3549, 3550], "Gin Lollipop['1 oz gin', '1 oz Southern Comfort® peach liqueur', '5 oz orange juice', '1 ozsparkling mineral water']": [5311, 5312], "Italian Margarita['1 oz amaretto almond liqueur', '2 oz sweet and sour mix', '1/2 oz Jose Cuervo® Especial gold tequila', '1/2 oz triple sec']": [6549, 6550], "Ma Wee Hen['1 1/2 oz Scotch whisky', '1/2 oz dry sherry', '1/2 oz orange juice', '1/2 oz lemon juice', '1 tsp grenadine syrup']": [7783, 7784], "Merete['2 cl Galliano® herbal liqueur', '2 cl Passoa® liqueur', '2 cl cranberry juice']": [8183, 8184], "Mozart Cocktail['1 1/2 oz anejo rum', '1/2 oz sweet vermouth', '1 tsp triple sec', '2 da

In [707]:
name_ing_ind_col_key = list(name_ing_ind_col.keys())
for i in range(len(name_ing_ind_col_key)):
    print(name_ing_ind_col[name_ing_ind_col_key[i]])
    for j in name_ing_ind_col[name_ing_ind_col_key[i]]:
        print(dricopy["Making"][j])
        print(dricopy["Glass"][j])
        print(dricopy["Type"][j])

[2527, 2528]
Shake ingredients over ice. Pour into glass.
Cocktail Glass
nan
Shake ingredients over ice. Pour into glass.
Cocktail Glass
nan
[2972, 2973]
Add Dr. Pepper and spiced rum, stir, then put Aftershock in and let it mix by itself.
Cup
nan
Add Dr. Pepper and spiced rum, stir, then put Aftershock in and let it mix by itself.
Cup
nan
[3549, 3550]
In a shaker half-filled with ice cubes, combine the gin and lemon juice. Shake well. Strain into a cocktail glass. Drop the grenadine into the center of the drink and float the port on the top.
Cocktail Glass
nan
Pour the gin and lemon juice into a cocktail shaker half-filled with ice cubes. Shake well. Strain into a cocktail glass. Drop the grenadine into the center of the drink and float port on top. Garnish with a maraschino cherry, and serve.
Cocktail Glass
nan
[5311, 5312]
Shake and strain into an ice-filled collins glass, and add water. Garnish with a slice of orange and a cherry, and serve.
Collins Glass
nan
Shake all ingredients 

같은 칵테일이 아닌 것: [9473, 9474], [15607, 15608]<br>
같은 칵테일이지만 설명 차이가 유의미한 수준이라서 하나를 선택해서 남겨야 하는 것: <br>
[3549, 3550]중 3550, [5311, 5312]중 5312, [6549, 6550]중 6549, [7783, 7784]중 7783, [8602, 8603]중 8603, [10408, 10409]중 10409, [11624, 11625]중 11625, [16123, 16124]중 16124<br>
중복이 아닌 것은 바리에이션 넘버 부여하기, 선택적으로 남겨야 하는 것 남기고 나머지 드랍하기

In [708]:
print(dricopy["Name"][9473])
print(dricopy["Name"][9474])
print(dricopy["Name"][15607])
print(dricopy["Name"][15608])

Pellston Hottie
Pellston Hottie
Banana Cantaloupe Smoothie
Banana Cantaloupe Smoothie


In [709]:
dricopy["Name"][9473] = "Pellston Hottie #1"
dricopy["Name"][9474] = "Pellston Hottie #2"
dricopy["Name"][15607] = "Banana Cantaloupe Smoothie #1"
dricopy["Name"][15608] = "Banana Cantaloupe Smoothie #2"

In [710]:
dricopy.drop(3549, inplace=True)
dricopy.drop(5311, inplace=True)
dricopy.drop(6550, inplace=True)
dricopy.drop(7784, inplace=True)
dricopy.drop(8602, inplace=True)
dricopy.drop(10408, inplace=True)
dricopy.drop(11624, inplace=True)
dricopy.drop(16123, inplace=True)

In [711]:
dricopy.reset_index(drop=True, inplace=True)

In [712]:
dricopy.drop_duplicates(subset=["Name", "Ingredients"], keep='first', inplace=True, ignore_index=True)

In [713]:
dri_2_dup = dricopy.duplicated(subset=["Name", "Ingredients"], keep=False)
len(dri_2_dup[dri_2_dup==True])

0

In [714]:
dricopy

Name                                        Ingredients  \
0            '61 Imperial  ['1/2 oz white rum', '1/2 oz vodka', '1/2 oz 1...   
1             .50 Caliber  ["2 oz Jack Daniel's® Tennessee whiskey", '2 o...   
2                     007  ['3 oz Stoli® Ohranj vodka', '1 1/2 ozfreshly ...   
3                 10 Deep  ['1.5 oz vodka', '1 oz Malibu® coconut rum', '...   
4      100 Miles per Hour  ['1/2 oz Wild Turkey® bourbon whiskey', '1/2 o...   
...                   ...                                                ...   
16795          Voodoo Tea  ['1 shot white rum', '1/2 shot Blue Curacao li...   
16796     Whiskey Punjabi  ['2 - 3 ozpremium whiskey', '8 - 12 ozunsweete...   
16797         Wild Beaver  ["3 oz Jack Daniel's® Tennessee whiskey", '3 o...   
16798     Xanthor's Pants  ['2 oz Jameson® Irish whiskey', "2 oz Bailey's...   
16799               Zorro  ['2 cl sambuca', "2 cl Bailey's® Irish cream",...   

                                                  Making  \
0      In a tall chimney glass or hurricane glass, fi...   
1      Pour the Jack Daniel's Tennessee whiskey, Jim ...   
2                      Serve on ice in a highball glass.   
3      All right start with 1/2 cup of ice mix the vo...   
4      Mix all ingredients together over ice in a hig...   
...                                                  ...   
16795  Pour rum and curacao into a highball glass. Fi...   
16796                                         Mix, stir.   
16797  Combine all ingredients in equal parts in a co...   
16798  Mix 1 part Scotch with 1 part Bailey's (approx...   
16799  Pour all ingredients into an irish coffee cup....   

                                               Link                Glass Type  \
0         http://www.drinksmixer.com/drink8428.html      Hurricane Glass  NaN   
1         http://www.drinksmixer.com/drink9993.html       Highball Glass  NaN   
2        http://www.drinksmixer.com/drink11090.html       Highball Glass  NaN   
3      http://www.drinksmixer.com/drinkuv17048.html                  Cup  NaN   
4        http://www.drinksmixer.com/drink10390.html       Highball Glass  NaN   
...                                             ...                  ...  ...   
16795     http://www.drinksmixer.com/drink6586.html       Highball Glass  NaN   
16796  http://www.drinksmixer.com/drinkvw15078.html  Old-Fashioned Glass  NaN   
16797     http://www.drinksmixer.com/drink8113.html         Beer Pilsner  NaN   
16798  http://www.drinksmixer.com/drinkrq15073.html           Coffee Mug  NaN   
16799     http://www.drinksmixer.com/drink6609.html     Irish Coffee Cup  NaN   

             Base  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
...           ...  
16795  Coffee&Tee  
16796  Coffee&Tee  
16797  Coffee&Tee  
16798  Coffee&Tee  
16799  Coffee&Tee  

[16800 rows x 7 columns]

4. 이름에서 중복 체크 후 바리에이션 넘버 부여하기

In [715]:
dri_name_dup = dricopy.duplicated(subset=["Name"], keep=False)
len(dri_name_dup[dri_name_dup==True])

295

In [716]:
dri_name_dup_ind = []
for i in range(len(dricopy)):
    if dri_name_dup[i]==True:
        dri_name_dup_ind.append(i)

In [717]:
name_unique = []
for i in dri_name_dup_ind:
    if dricopy["Name"][i] not in name_unique:
        name_unique.append(dricopy["Name"][i])
print(len(name_unique))

146


바리에이션 넘버 부여

바리에이션 넘버가 이미 부여된 상태에서 이름이 같은 칵테일이 있는지 확인

In [718]:
variation_dup_ind = []
for i in dri_name_dup_ind:
    if '#' in dricopy["Name"][i]:
        print("here")
        print(dricopy["Name"][i])
        variation_dup_ind.append(i)

here
Almond Joy #2
here
Banana Colada #2
here
Irish Cream #2
here
Irish Cream #2
here
Red Devil #2
here
Spanish Coffee #2
here
Almond Joy #2
here
Apple Pie Shot #3
here
Apple Pie Shot #3
here
Red Devil #2
here
Banana Colada #2
here
Spanish Coffee #2


해당 칵테일들은 일단 제외하고 바리에이션 넘버 부여하기

In [719]:
name_dup_sub_vari_ind = [x for x in dri_name_dup_ind if x not in variation_dup_ind]

In [720]:
not_vari_name_unique = []
for i in name_dup_sub_vari_ind:
    if dricopy["Name"][i] not in not_vari_name_unique:
        not_vari_name_unique.append(dricopy["Name"][i])
print(len(not_vari_name_unique))

140


In [721]:
for name in not_vari_name_unique:
    dri_same_names_ind = []
    for j in range(len(dricopy)):
        if dricopy["Name"][j]==name:
            dri_same_names_ind.append(j)
    for k in range(len(dri_same_names_ind)):
         dricopy["Name"][dri_same_names_ind[k]] = "{} #{}".format(name, k+1)

다시 이름 중복 확인

In [722]:
vari_name_dup = dricopy.duplicated(subset=["Name"], keep=False)
len(vari_name_dup[vari_name_dup==True])

61

In [723]:
vari_name_dup_ind = []
for i in range(len(dricopy)):
    if vari_name_dup[i]==True:
        vari_name_dup_ind.append(i)

In [724]:
vari_name_unique = []
for i in vari_name_dup_ind:
    if dricopy["Name"][i] not in vari_name_unique:
        vari_name_unique.append(dricopy["Name"][i])
print(len(vari_name_unique))

30


In [725]:
print(vari_name_unique)

['Almond Joy #2', 'Banana Colada #2', 'Cherry Blossom #2', 'Cherry Bomb #2', 'Cotton Candy #2', 'Cranberry Kiss #2', 'Espresso Martini #2', 'Fallen Angel #2', 'Gingerbread Man #2', 'Golden Shower #2', 'Irish Cream #2', 'Lava Lamp #2', 'Mango Tango #2', 'Metropolitan #2', 'Panty Dropper #2', 'Red Devil #2', 'Spanish Coffee #2', 'Strawberry Dream #2', 'Velvet Hammer #2', 'Apple Pie Shot #2', 'Apple Pie Shot #3', 'Black Death #2', 'Black Tooth #2', 'Blackjack #2', 'Cherry Coke #2', 'Purple Haze #2', 'Spiced Apple #2', 'Watermelon #2', 'Banana Punch #2', 'Irish Cream Liqueur #2']


In [726]:
root_name_unique = []
for i in range(len(vari_name_unique)):
    root_name_unique.append(vari_name_unique[i].split("#")[0].strip())
root_name_unique = list(set(root_name_unique))
print(root_name_unique)

['Irish Cream Liqueur', 'Espresso Martini', 'Black Death', 'Almond Joy', 'Cherry Blossom', 'Panty Dropper', 'Purple Haze', 'Banana Colada', 'Banana Punch', 'Blackjack', 'Black Tooth', 'Lava Lamp', 'Fallen Angel', 'Gingerbread Man', 'Cherry Bomb', 'Spanish Coffee', 'Spiced Apple', 'Cherry Coke', 'Golden Shower', 'Red Devil', 'Watermelon', 'Velvet Hammer', 'Cotton Candy', 'Strawberry Dream', 'Cranberry Kiss', 'Mango Tango', 'Apple Pie Shot', 'Metropolitan', 'Irish Cream']


In [727]:
dricopycopy = dricopy.copy()

겹치지 않는 새로운 바리에이션 넘버 부여해주기

In [728]:
for name in root_name_unique:
    vari_counts = []
    vari_ind = []
    for i in range(len(dricopycopy)):
        if "#" in dricopycopy["Name"][i]:
            if dricopycopy["Name"][i].split("#")[0].strip()==name:
                vari_counts.append(int(dricopycopy["Name"][i].split("#")[1].strip()))
                vari_ind.append(i)
    vari_counts = list(range(len(vari_counts)))
    for j in vari_ind:
        dricopycopy["Name"][j] = "{} #{}".format(name, j+1)

다시 중복 확인

In [729]:
name_dup = dricopycopy.duplicated(subset=["Name"], keep=False)
len(name_dup[name_dup==True])

0

In [730]:
dricopycopy

Name                                        Ingredients  \
0            '61 Imperial  ['1/2 oz white rum', '1/2 oz vodka', '1/2 oz 1...   
1             .50 Caliber  ["2 oz Jack Daniel's® Tennessee whiskey", '2 o...   
2                     007  ['3 oz Stoli® Ohranj vodka', '1 1/2 ozfreshly ...   
3                 10 Deep  ['1.5 oz vodka', '1 oz Malibu® coconut rum', '...   
4      100 Miles per Hour  ['1/2 oz Wild Turkey® bourbon whiskey', '1/2 o...   
...                   ...                                                ...   
16795          Voodoo Tea  ['1 shot white rum', '1/2 shot Blue Curacao li...   
16796     Whiskey Punjabi  ['2 - 3 ozpremium whiskey', '8 - 12 ozunsweete...   
16797         Wild Beaver  ["3 oz Jack Daniel's® Tennessee whiskey", '3 o...   
16798     Xanthor's Pants  ['2 oz Jameson® Irish whiskey', "2 oz Bailey's...   
16799               Zorro  ['2 cl sambuca', "2 cl Bailey's® Irish cream",...   

                                                  Making  \
0      In a tall chimney glass or hurricane glass, fi...   
1      Pour the Jack Daniel's Tennessee whiskey, Jim ...   
2                      Serve on ice in a highball glass.   
3      All right start with 1/2 cup of ice mix the vo...   
4      Mix all ingredients together over ice in a hig...   
...                                                  ...   
16795  Pour rum and curacao into a highball glass. Fi...   
16796                                         Mix, stir.   
16797  Combine all ingredients in equal parts in a co...   
16798  Mix 1 part Scotch with 1 part Bailey's (approx...   
16799  Pour all ingredients into an irish coffee cup....   

                                               Link                Glass Type  \
0         http://www.drinksmixer.com/drink8428.html      Hurricane Glass  NaN   
1         http://www.drinksmixer.com/drink9993.html       Highball Glass  NaN   
2        http://www.drinksmixer.com/drink11090.html       Highball Glass  NaN   
3      http://www.drinksmixer.com/drinkuv17048.html                  Cup  NaN   
4        http://www.drinksmixer.com/drink10390.html       Highball Glass  NaN   
...                                             ...                  ...  ...   
16795     http://www.drinksmixer.com/drink6586.html       Highball Glass  NaN   
16796  http://www.drinksmixer.com/drinkvw15078.html  Old-Fashioned Glass  NaN   
16797     http://www.drinksmixer.com/drink8113.html         Beer Pilsner  NaN   
16798  http://www.drinksmixer.com/drinkrq15073.html           Coffee Mug  NaN   
16799     http://www.drinksmixer.com/drink6609.html     Irish Coffee Cup  NaN   

             Base  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
...           ...  
16795  Coffee&Tee  
16796  Coffee&Tee  
16797  Coffee&Tee  
16798  Coffee&Tee  
16799  Coffee&Tee  

[16800 rows x 7 columns]

### Ingredients 형식 맞추기

양 끝의 '['와 ']' 떼주기

In [731]:
for i in range(len(dricopycopy)):
    ingredients = dricopycopy["Ingredients"][i][1:-1]
    dricopycopy["Ingredients"][i] = ingredients

In [732]:
# 확인
print(dricopycopy["Ingredients"][1])

"2 oz Jack Daniel's® Tennessee whiskey", '2 oz Jim Beam® bourbon whiskey', '2 oz dry gin', '2 oz Absolut® vodka', '1 oz Squirt® citrus soda'


"를 '로 바꾸기

In [733]:
for i in range(len(dricopycopy["Ingredients"])):
    dricopycopy["Ingredients"][i] = dricopycopy["Ingredients"][i].replace('"', "'")

In [734]:
# 확인
print(dricopycopy["Ingredients"][1])

'2 oz Jack Daniel's® Tennessee whiskey', '2 oz Jim Beam® bourbon whiskey', '2 oz dry gin', '2 oz Absolut® vodka', '1 oz Squirt® citrus soda'


양 끝의 ' 떼주기

In [735]:
for i in range(len(dricopycopy)):
    ingredients = dricopycopy["Ingredients"][i][1:-1]
    dricopycopy["Ingredients"][i] = ingredients

In [736]:
# 확인
print(dricopycopy["Ingredients"][1])

2 oz Jack Daniel's® Tennessee whiskey', '2 oz Jim Beam® bourbon whiskey', '2 oz dry gin', '2 oz Absolut® vodka', '1 oz Squirt® citrus soda


단위를 'shot', 'shots', 'ml', 'cl'에서 'oz'로 바꾸기

In [737]:
from fractions import Fraction

In [738]:
def is_num(n):
    try:
        float(n)
    except:
        return False
    return True

In [739]:
for i in range(len(dricopycopy["Ingredients"])):
    ingredients = dricopycopy["Ingredients"][i].lower().split("', '")
    for j in range(len(ingredients)):
        detail = ingredients[j].strip().split(" ")
        if len(detail)>=3:
            if is_num(detail[0]) and is_num(detail[1]):
                frac1 = Fraction(detail[0]).limit_denominator(100)
                frac2 = Fraction(detail[1]).limit_denominator(100)
                if detail[2]=="shot":
                    detail[2] = "oz"
                elif detail[2]=="shots":
                    detail[2] = "oz"
                elif detail[2]=="ml":
                    if frac1+frac2>=9:
                        oz = Fraction(Fraction(frac1+frac2), 30).limit_denominator(100)
                        rest = detail[3:]
                        new_volume = [str(oz), "oz"]
                        new_volume.extend(rest)
                        detail = new_volume
                elif detail[2]=="cl":
                    if frac1+frac2>=0.9:
                        oz = Fraction(Fraction(frac1+frac2), 3).limit_denominator(100)
                        rest = detail[3:]
                        new_volume = [str(oz), "oz"]
                        new_volume.extend(rest)
                        detail = new_volume
                else:
                    pass
            else:
                if is_num(detail[0]):
                    frac = Fraction(detail[0]).limit_denominator(100)
                    if detail[1]=="shot":
                        detail[1] = "oz"
                    elif detail[1]=="shots":
                        detail[1] = "oz"
                    elif detail[1]=="ml":
                        if frac>=9:
                            oz = Fraction(Fraction(frac), 30).limit_denominator(100)
                            rest = detail[2:]
                            new_volume = [str(oz), "oz"]
                            new_volume.extend(rest)
                            detail = new_volume
                    elif detail[1]=="cl":
                        if frac>=0.9:
                            oz = Fraction(Fraction(frac), 3).limit_denominator(100)
                            rest = detail[2:]
                            new_volume = [str(oz), "oz"]
                            new_volume.extend(rest)
                            detail = new_volume
                    else:
                        pass
                else:
                    pass
        ingredients[j] = " ".join(detail)
    new_ing = ", ".join(ingredients)
    dricopycopy["Ingredients"][i] = new_ing

In [740]:
# 확인
dricopycopy

Name                                        Ingredients  \
0            '61 Imperial  1/2 oz white rum, 1/2 oz vodka, 1/2 oz 151 pro...   
1             .50 Caliber  2 oz jack daniel's® tennessee whiskey, 2 oz ji...   
2                     007  3 oz stoli® ohranj vodka, 1 1/2 ozfreshly sque...   
3                 10 Deep  1.5 oz vodka, 1 oz malibu® coconut rum, 1/2 cu...   
4      100 Miles per Hour  1/2 oz wild turkey® bourbon whiskey, 1/2 oz ba...   
...                   ...                                                ...   
16795          Voodoo Tea  1 oz white rum, 1/2 shot blue curacao liqueur,...   
16796     Whiskey Punjabi  2 - 3 ozpremium whiskey, 8 - 12 ozunsweetened ...   
16797         Wild Beaver  3 oz jack daniel's® tennessee whiskey, 3 oz am...   
16798     Xanthor's Pants  2 oz jameson® irish whiskey, 2 oz bailey's® ir...   
16799               Zorro  2/3 oz sambuca, 2/3 oz bailey's® irish cream, ...   

                                                  Making  \
0      In a tall chimney glass or hurricane glass, fi...   
1      Pour the Jack Daniel's Tennessee whiskey, Jim ...   
2                      Serve on ice in a highball glass.   
3      All right start with 1/2 cup of ice mix the vo...   
4      Mix all ingredients together over ice in a hig...   
...                                                  ...   
16795  Pour rum and curacao into a highball glass. Fi...   
16796                                         Mix, stir.   
16797  Combine all ingredients in equal parts in a co...   
16798  Mix 1 part Scotch with 1 part Bailey's (approx...   
16799  Pour all ingredients into an irish coffee cup....   

                                               Link                Glass Type  \
0         http://www.drinksmixer.com/drink8428.html      Hurricane Glass  NaN   
1         http://www.drinksmixer.com/drink9993.html       Highball Glass  NaN   
2        http://www.drinksmixer.com/drink11090.html       Highball Glass  NaN   
3      http://www.drinksmixer.com/drinkuv17048.html                  Cup  NaN   
4        http://www.drinksmixer.com/drink10390.html       Highball Glass  NaN   
...                                             ...                  ...  ...   
16795     http://www.drinksmixer.com/drink6586.html       Highball Glass  NaN   
16796  http://www.drinksmixer.com/drinkvw15078.html  Old-Fashioned Glass  NaN   
16797     http://www.drinksmixer.com/drink8113.html         Beer Pilsner  NaN   
16798  http://www.drinksmixer.com/drinkrq15073.html           Coffee Mug  NaN   
16799     http://www.drinksmixer.com/drink6609.html     Irish Coffee Cup  NaN   

             Base  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
...           ...  
16795  Coffee&Tee  
16796  Coffee&Tee  
16797  Coffee&Tee  
16798  Coffee&Tee  
16799  Coffee&Tee  

[16800 rows x 7 columns]

### Making 형식 맞추기

In [741]:
for i in range(len(dricopycopy)):
    making = dricopycopy["Making"][i].replace("\n", " ")
    making.replace("  ", " ")
    dricopycopy["Making"][i] = making

### Glass 형식 맞추기

In [742]:
dricopycopy["Glass"].unique()

array(['Hurricane Glass', 'Highball Glass', 'Cup', 'Collins Glass',
       'Cocktail Glass', 'Old-Fashioned Glass', 'Margarita Glass', nan,
       'Beer Mug', 'Sour Glass', 'White Wine Glass', 'Beer Pilsner',
       'Champagne Flute', 'Pina Colada Glass', 'Wine Goblet', 'Mason Jar',
       'Cordial Glass', 'Coffee Mug', 'Shot Glass', 'Champagne Saucer',
       'Bottle', 'Pitcher', 'Irish Coffee Cup', 'Cooler',
       'Brandy Snifter', 'Red Wine Glass', 'Mug', 'Pousse Cafe Glass',
       'Punch Bowl', 'Parfait Glass', 'Pint glass', 'Sherry Glass',
       'Champagne Tulip', 'Coupe glass', 'Jug', 'Test tube',
       'Aperitif Glass', 'Bucket', 'Saucepan'], dtype=object)

In [743]:
for i in range(len(dricopycopy)):
    if str(dricopycopy["Glass"][i])=='nan':
        dricopycopy["Glass"][i] = None

In [744]:
for i in range(len(dricopycopy)):
    if dricopycopy["Glass"][i]=='Old-Fashioned Glass':
        dricopycopy["Glass"][i] = "Old Fashioned Glass"
    if dricopycopy["Glass"][i]=='Beer Pilsner':
        dricopycopy["Glass"][i] = "Pilsner Glass"
    if dricopycopy["Glass"][i]=='Champagne Flute':
        dricopycopy["Glass"][i] = "Flute Glass"
    if dricopycopy["Glass"][i]=='Champagne Saucer':
        dricopycopy["Glass"][i] = "Saucer Glass"
    if dricopycopy["Glass"][i]=='Pitcher':
        dricopycopy["Glass"][i] = "Pitcher Glass"
    if dricopycopy["Glass"][i]=='Cooler':
        dricopycopy["Glass"][i] = "Cooler Glass"
    if dricopycopy["Glass"][i]=='Brandy Snifter':
        dricopycopy["Glass"][i] = "Snifter Glass"
    if dricopycopy["Glass"][i]=='Pint glass':
        dricopycopy["Glass"][i] = "Pint Glass"
    if dricopycopy["Glass"][i]=='Champagne Tulip':
        dricopycopy["Glass"][i] = "Tulip Glass"
    if dricopycopy["Glass"][i]=='Coupe glass':
        dricopycopy["Glass"][i] = "Coupe Glass"
    if dricopycopy["Glass"][i]=='Test tube':
        dricopycopy["Glass"][i] = "Test Tube"
    if dricopycopy["Glass"][i]=='Saucepan':
        dricopycopy["Glass"][i] = None

In [745]:
dricopycopy["Glass"].unique()

array(['Hurricane Glass', 'Highball Glass', 'Cup', 'Collins Glass',
       'Cocktail Glass', 'Old Fashioned Glass', 'Margarita Glass', None,
       'Beer Mug', 'Sour Glass', 'White Wine Glass', 'Pilsner Glass',
       'Flute Glass', 'Pina Colada Glass', 'Wine Goblet', 'Mason Jar',
       'Cordial Glass', 'Coffee Mug', 'Shot Glass', 'Saucer Glass',
       'Bottle', 'Pitcher Glass', 'Irish Coffee Cup', 'Cooler Glass',
       'Snifter Glass', 'Red Wine Glass', 'Mug', 'Pousse Cafe Glass',
       'Punch Bowl', 'Parfait Glass', 'Pint Glass', 'Sherry Glass',
       'Tulip Glass', 'Coupe Glass', 'Jug', 'Test Tube', 'Aperitif Glass',
       'Bucket'], dtype=object)

### Garnish 추출

In [746]:
garnish_list = [None]*16800
for i in range(len(dricopycopy)):
    making = dricopycopy["Making"][i].lower()
    if "garnish with" in making:
        if making.count("garnish with")>1:
            print("many garnish: ", i)
        else:
            start = making.index("garnish with")+12
            try:
                try:
                    end = making.index(".", start)
                except:
                    end = making.index(",", start)
            except:
                end = None
            if end != None:
                garnish = making[start:end]
                garnish_list[i] = garnish
            else:
                garnish = making[start:]
                garnish_list[i] = garnish

many garnish:  9738
many garnish:  15842


In [747]:
print(dricopycopy["Making"][9738])
print(dricopycopy["Making"][15842])

Blend together egg whites and white sugar in blender for 15 seconds. Add Pisco and lime juice and blend for 15 seconds.  Put crushed ice into shaker and add Pisco sour mix to about 2/3 full. Shake and strain into small glasses. Garnish with a sprinkle of brown sugar. You can also garnish with a drizzle of browned, melted butter.  Repeat shaker process with rest of mix, only adding more crushed ice as needed.  Makes about 16 servings - enough for 8 people, 2 servings each.
Combine all ingredients, except ice cubes & mint, in a blender container and blend on high speed until frothy. Pour into 6 tall glasses filled with ice cubes. Garnish with mint sprigs, if desired. Thicker Variation: Omit ice cubes. Freeze the pineapple juice in ice-cube trays, blend these cubes with the other ingredients and garnish with mint.


In [748]:
garnish_list[9738] = "a sprinkle of brown sugar or a drizzle of browned, melted butter"
garnish_list[15842] = "mint sprigs"

In [749]:
dricopycopy["Garnish"] = garnish_list
dricopycopy.head()

Name                                        Ingredients  \
0        '61 Imperial  1/2 oz white rum, 1/2 oz vodka, 1/2 oz 151 pro...   
1         .50 Caliber  2 oz jack daniel's® tennessee whiskey, 2 oz ji...   
2                 007  3 oz stoli® ohranj vodka, 1 1/2 ozfreshly sque...   
3             10 Deep  1.5 oz vodka, 1 oz malibu® coconut rum, 1/2 cu...   
4  100 Miles per Hour  1/2 oz wild turkey® bourbon whiskey, 1/2 oz ba...   

                                              Making  \
0  In a tall chimney glass or hurricane glass, fi...   
1  Pour the Jack Daniel's Tennessee whiskey, Jim ...   
2                  Serve on ice in a highball glass.   
3  All right start with 1/2 cup of ice mix the vo...   
4  Mix all ingredients together over ice in a hig...   

                                           Link            Glass Type Base  \
0     http://www.drinksmixer.com/drink8428.html  Hurricane Glass  NaN  NaN   
1     http://www.drinksmixer.com/drink9993.html   Highball Glass  NaN  NaN   
2    http://www.drinksmixer.com/drink11090.html   Highball Glass  NaN  NaN   
3  http://www.drinksmixer.com/drinkuv17048.html              Cup  NaN  NaN   
4    http://www.drinksmixer.com/drink10390.html   Highball Glass  NaN  NaN   

                                             Garnish  
0   a pineapple slice and a maraschino cherry, an...  
1                                               None  
2                                               None  
3                                               None  
4                                               None

### 추가로 스크랩해놓은 Base, Type 데이터 부여하기

In [750]:
dricopycopy["Base"].unique()

array([nan, '', 'Coffee&Tee', 'Beer&Ale'], dtype=object)

In [751]:
dricopycopy["Type"].unique()

array([nan, 'Shooter', '', 'Punch', 'Nonalcoholic'], dtype=object)

In [752]:
dri_base_brandy = pd.read_csv("//Drink_base_brandy.csv")
brandy_list = list(dri_base_brandy["Name"])
dri_base_champagne = pd.read_csv("//Drink_base_champagne.csv")
champagne_list = list(dri_base_champagne["Name"])
dri_base_cognac = pd.read_csv("//Drink_base_cognac.csv")
cognac_list = list(dri_base_cognac["Name"])
dri_base_gin = pd.read_csv("//Drink_base_gin.csv")
gin_list = list(dri_base_gin["Name"])
dri_base_liquor = pd.read_csv("//Drink_base_liqueur.csv")
liquor_list = list(dri_base_liquor["Name"])
dri_base_rum = pd.read_csv("//Drink_base_rum.csv")
rum_list = list(dri_base_rum["Name"])
dri_base_tequila = pd.read_csv("//Drink_base_tequila.csv")
tequila_list = list(dri_base_tequila["Name"])
dri_base_vodka = pd.read_csv("//Drink_base_vodka.csv")
vodka_list = list(dri_base_vodka["Name"])
dri_base_whisky = pd.read_csv("//Drink_base_whisky.csv")
whisky_list = list(dri_base_whisky["Name"])
dri_base_wine = pd.read_csv("//Drink_base_wine.csv")
wine_list = list(dri_base_wine["Name"])

dri_type_creamy = pd.read_csv("//Drink_type_creamy.csv")
creamy_list = list(dri_type_creamy["Name"])
dri_type_frozen = pd.read_csv("//Drink_type_frozen.csv")
frozen_list = list(dri_type_frozen["Name"])
dri_type_hot = pd.read_csv("//Drink_type_hot.csv")
hot_list = list(dri_type_hot["Name"])
dri_type_long = pd.read_csv("//Drink_type_long.csv")
long_list = list(dri_type_long["Name"])
dri_type_short = pd.read_csv("//Drink_type_short.csv")
short_list = list(dri_type_short["Name"])
dri_type_tropical = pd.read_csv("//Drink_type_tropical.csv")
tropical_list = list(dri_type_tropical["Name"])

In [767]:
base_list = [None]*16800
type_list = [None]*16800
for i in range(len(dricopy)):
    bases = []
    types = []
    if dricopycopy["Base"][i]=='Beer&Ale':
        if 'beer'in dricopycopy["Ingredients"][i].lower():
            bases.append("Beer")
        elif ' ale ' in dricopycopy["Ingredients"][i].lower():
            bases.append("Ale")
        else:
            bases.append("Beer & Ale")
    if dricopycopy["Base"][i]=='Coffee&Tee':
        bases.append("Coffee & Tee")
        
    if dricopycopy["Name"][i] in brandy_list:
        bases.append("Brandy")
    if dricopycopy["Name"][i] in champagne_list:
        bases.append("Champagne")
    if dricopycopy["Name"][i] in cognac_list:
        bases.append("Cognac")
    if dricopycopy["Name"][i] in gin_list:
        bases.append("Gin")
    if dricopycopy["Name"][i] in liquor_list:
        bases.append("Liquor")
    if dricopycopy["Name"][i] in rum_list:
        bases.append("Rum")
    if dricopycopy["Name"][i] in tequila_list:
        bases.append("Tequila")
    if dricopycopy["Name"][i] in vodka_list:
        bases.append("Vodka")
    if dricopycopy["Name"][i] in whisky_list:
        bases.append("Whisky")
    if dricopycopy["Name"][i] in wine_list:
        bases.append("Wine")
        
    if "vodka" in dricopycopy["Ingredients"][i].lower():
        bases.append("Vodka")
    if "rum" in dricopycopy["Ingredients"][i].lower():
        bases.append("Rum")
    if "whisky" in dricopycopy["Ingredients"][i].lower() or "whiskey" in dricopycopy["Ingredients"][i].lower():
        bases.append("Whisky")
    if "tequila" in dricopycopy["Ingredients"][i].lower() or "tekila" in dricopycopy["Ingredients"][i].lower() or "teqila" in dricopycopy["Ingredients"][i].lower():
        bases.append("Tequila")
    if "mezcal" in dricopycopy["Ingredients"][i].lower():
        bases.append("Mezcal")
    if "gin" in dricopycopy["Ingredients"][i].lower():
        bases.append("Gin")
    if "cognac" in dricopycopy["Ingredients"][i].lower():
        bases.append("Cognac")
    if "brandy" in dricopycopy["Ingredients"][i].lower():
        bases.append("Brandy")
    if "wine" in dricopycopy["Ingredients"][i].lower():
        bases.append("Wine")
    if "liquor" in dricopycopy["Ingredients"][i].lower() or "liqueur" in dricopycopy["Ingredients"][i].lower() or "liqour" in dricopycopy["Ingredients"][i].lower() or "liquour" in dricopycopy["Ingredients"][i].lower():
        bases.append("Liquor")
    if "champagne" in dricopycopy["Ingredients"][i].lower():
        bases.append("Champagne")
        
    if dricopycopy["Type"][i]=='Nonalcoholic':
        types.append("Non-Alcoholic")
    if dricopycopy["Name"][i] in creamy_list:
        types.append("Creamy")
    if dricopycopy["Name"][i] in frozen_list:
        types.append("Frozen")
    if dricopycopy["Name"][i] in hot_list:
        types.append("Hot")
    if dricopycopy["Name"][i] in long_list:
        types.append("Long")
    if dricopycopy["Name"][i] in short_list:
        types.append("Short")
    if dricopycopy["Name"][i] in tropical_list:
        types.append("Tropical")
    
    bases = list(set(bases))
    if "Mezcal" in bases and "Tequila" in bases:
        bases.remove("Mezcal")
    if "Brandy" in bases and "Cognac" in bases:
        bases.remove("Brandy")
    if "Beer" in bases and "Ale" in bases:
        print(i)
#        bases.remove("Beer")
#        bases.remove("Ale")
#        bases.append("")
    
    base_list[i] = ", ".join(bases)
    type_list[i] = ", ".join(types)

In [768]:
# 확인
dricopycopy["Base"] = base_list
dricopycopy["Type"] = type_list
dricopycopy.head()

Name                                        Ingredients  \
0        '61 Imperial  1/2 oz white rum, 1/2 oz vodka, 1/2 oz 151 pro...   
1         .50 Caliber  2 oz jack daniel's® tennessee whiskey, 2 oz ji...   
2                 007  3 oz stoli® ohranj vodka, 1 1/2 ozfreshly sque...   
3             10 Deep  1.5 oz vodka, 1 oz malibu® coconut rum, 1/2 cu...   
4  100 Miles per Hour  1/2 oz wild turkey® bourbon whiskey, 1/2 oz ba...   

                                              Making  \
0  In a tall chimney glass or hurricane glass, fi...   
1  Pour the Jack Daniel's Tennessee whiskey, Jim ...   
2                  Serve on ice in a highball glass.   
3  All right start with 1/2 cup of ice mix the vo...   
4  Mix all ingredients together over ice in a hig...   

                                           Link            Glass  Type  \
0     http://www.drinksmixer.com/drink8428.html  Hurricane Glass  Long   
1     http://www.drinksmixer.com/drink9993.html   Highball Glass  Long   
2    http://www.drinksmixer.com/drink11090.html   Highball Glass  Long   
3  http://www.drinksmixer.com/drinkuv17048.html              Cup         
4    http://www.drinksmixer.com/drink10390.html   Highball Glass  Long   

                  Base                                            Garnish  \
0   Rum, Vodka, Liquor   a pineapple slice and a maraschino cherry, an...   
1   Vodka, Whisky, Gin                                               None   
2                Vodka                                               None   
3           Rum, Vodka                                               None   
4  Rum, Whisky, Liquor                                               None   

        Method  
0  Stir, Float  
1         None  
2         None  
3         None  
4        Shake

비정상 Type 제거

In [770]:
dricopycopy["Type"].unique()

array(['Long', '', 'Long, Short', 'Short', 'Creamy, Short',
       'Long, Tropical', 'Creamy, Long', 'Frozen, Long',
       'Frozen, Long, Short', 'Hot, Short', 'Frozen, Short', 'Frozen',
       'Hot', 'Creamy, Long, Short', 'Creamy, Frozen, Long, Short',
       'Creamy, Frozen, Long', 'Short, Tropical', 'Creamy, Frozen, Short',
       'Frozen, Short, Tropical', 'Tropical', 'Hot, Long',
       'Frozen, Long, Tropical', 'Frozen, Long, Short, Tropical',
       'Hot, Long, Short'], dtype=object)

In [771]:
for i in range(len(dricopy)):
    if dricopycopy["Type"][i]=='Long, Short':
        dricopycopy["Type"][i] = None
    if dricopycopy["Type"][i]=='Frozen, Long, Short':
        dricopycopy["Type"][i] = None
    if dricopycopy["Type"][i]=='Creamy, Long, Short':
        dricopycopy["Type"][i] = None
    if dricopycopy["Type"][i]=='Creamy, Frozen, Long, Short':
        dricopycopy["Type"][i] = None
    if dricopycopy["Type"][i]=='Frozen, Long, Short, Tropical':
        dricopycopy["Type"][i] = None
    if dricopycopy["Type"][i]=='Hot, Long, Short':
        dricopycopy["Type"][i] = None

In [772]:
dricopycopy.head()

Name                                        Ingredients  \
0        '61 Imperial  1/2 oz white rum, 1/2 oz vodka, 1/2 oz 151 pro...   
1         .50 Caliber  2 oz jack daniel's® tennessee whiskey, 2 oz ji...   
2                 007  3 oz stoli® ohranj vodka, 1 1/2 ozfreshly sque...   
3             10 Deep  1.5 oz vodka, 1 oz malibu® coconut rum, 1/2 cu...   
4  100 Miles per Hour  1/2 oz wild turkey® bourbon whiskey, 1/2 oz ba...   

                                              Making  \
0  In a tall chimney glass or hurricane glass, fi...   
1  Pour the Jack Daniel's Tennessee whiskey, Jim ...   
2                  Serve on ice in a highball glass.   
3  All right start with 1/2 cup of ice mix the vo...   
4  Mix all ingredients together over ice in a hig...   

                                           Link            Glass  Type  \
0     http://www.drinksmixer.com/drink8428.html  Hurricane Glass  Long   
1     http://www.drinksmixer.com/drink9993.html   Highball Glass  Long   
2    http://www.drinksmixer.com/drink11090.html   Highball Glass  Long   
3  http://www.drinksmixer.com/drinkuv17048.html              Cup         
4    http://www.drinksmixer.com/drink10390.html   Highball Glass  Long   

                  Base                                            Garnish  \
0   Rum, Vodka, Liquor   a pineapple slice and a maraschino cherry, an...   
1   Vodka, Whisky, Gin                                               None   
2                Vodka                                               None   
3           Rum, Vodka                                               None   
4  Rum, Whisky, Liquor                                               None   

        Method  
0  Stir, Float  
1         None  
2         None  
3         None  
4        Shake

### Method 추출

In [773]:
method_list = [None]*16800
for i in range(len(dricopycopy)):
    methods = []
    if "shake" in dricopycopy["Making"][i].lower() or "shkae" in dricopycopy["Making"][i].lower():
        methods.append("Shake")
    if "stir" in dricopycopy["Making"][i].lower():
            methods.append("Stir")
    elif "mixing glass" in dricopycopy["Making"][i].lower():
        methods.append("Stir")
    elif "mix over" in dricopycopy["Making"][i].lower():
        methods.append("Stir")
    elif "mix well" in dricopycopy["Making"][i].lower():
        methods.append("Stir")
    if "build" in dricopycopy["Making"][i].lower():
        methods.append("Build")
    if "combine" in dricopycopy["Making"][i].lower():
        methods.append("Build")
    if "float" in dricopycopy["Making"][i].lower() or "layer" in dricopycopy["Making"][i].lower():
        methods.append("Float")
    if "blend" in dricopycopy["Making"][i].lower():
        methods.append("Blend")
    if "strain" in dricopycopy["Making"][i].lower():
        methods.append("Strain")
    methods = list(set(methods))
    if methods==[]:
        pass
    else:
        method_list[i]=", ".join(methods)

Method 속성 부여하고 확인하기

In [774]:
dricopycopy["Method"] = method_list
dricopycopy.head()

Name                                        Ingredients  \
0        '61 Imperial  1/2 oz white rum, 1/2 oz vodka, 1/2 oz 151 pro...   
1         .50 Caliber  2 oz jack daniel's® tennessee whiskey, 2 oz ji...   
2                 007  3 oz stoli® ohranj vodka, 1 1/2 ozfreshly sque...   
3             10 Deep  1.5 oz vodka, 1 oz malibu® coconut rum, 1/2 cu...   
4  100 Miles per Hour  1/2 oz wild turkey® bourbon whiskey, 1/2 oz ba...   

                                              Making  \
0  In a tall chimney glass or hurricane glass, fi...   
1  Pour the Jack Daniel's Tennessee whiskey, Jim ...   
2                  Serve on ice in a highball glass.   
3  All right start with 1/2 cup of ice mix the vo...   
4  Mix all ingredients together over ice in a hig...   

                                           Link            Glass  Type  \
0     http://www.drinksmixer.com/drink8428.html  Hurricane Glass  Long   
1     http://www.drinksmixer.com/drink9993.html   Highball Glass  Long   
2    http://www.drinksmixer.com/drink11090.html   Highball Glass  Long   
3  http://www.drinksmixer.com/drinkuv17048.html              Cup         
4    http://www.drinksmixer.com/drink10390.html   Highball Glass  Long   

                  Base                                            Garnish  \
0   Rum, Vodka, Liquor   a pineapple slice and a maraschino cherry, an...   
1   Vodka, Whisky, Gin                                               None   
2                Vodka                                               None   
3           Rum, Vodka                                               None   
4  Rum, Whisky, Liquor                                               None   

        Method  
0  Stir, Float  
1         None  
2         None  
3         None  
4        Shake

빈 Flavor 열 추가하기

In [775]:
dricopycopy["Flavor"]=None
dricopycopy.head()

Name                                        Ingredients  \
0        '61 Imperial  1/2 oz white rum, 1/2 oz vodka, 1/2 oz 151 pro...   
1         .50 Caliber  2 oz jack daniel's® tennessee whiskey, 2 oz ji...   
2                 007  3 oz stoli® ohranj vodka, 1 1/2 ozfreshly sque...   
3             10 Deep  1.5 oz vodka, 1 oz malibu® coconut rum, 1/2 cu...   
4  100 Miles per Hour  1/2 oz wild turkey® bourbon whiskey, 1/2 oz ba...   

                                              Making  \
0  In a tall chimney glass or hurricane glass, fi...   
1  Pour the Jack Daniel's Tennessee whiskey, Jim ...   
2                  Serve on ice in a highball glass.   
3  All right start with 1/2 cup of ice mix the vo...   
4  Mix all ingredients together over ice in a hig...   

                                           Link            Glass  Type  \
0     http://www.drinksmixer.com/drink8428.html  Hurricane Glass  Long   
1     http://www.drinksmixer.com/drink9993.html   Highball Glass  Long   
2    http://www.drinksmixer.com/drink11090.html   Highball Glass  Long   
3  http://www.drinksmixer.com/drinkuv17048.html              Cup         
4    http://www.drinksmixer.com/drink10390.html   Highball Glass  Long   

                  Base                                            Garnish  \
0   Rum, Vodka, Liquor   a pineapple slice and a maraschino cherry, an...   
1   Vodka, Whisky, Gin                                               None   
2                Vodka                                               None   
3           Rum, Vodka                                               None   
4  Rum, Whisky, Liquor                                               None   

        Method Flavor  
0  Stir, Float   None  
1         None   None  
2         None   None  
3         None   None  
4        Shake   None

카테고리 순서 형식에 맞게 정렬하고 확인하기

In [776]:
dricopycopy = dricopycopy[["Name", "Ingredients", "Making", "Link", "Glass", "Garnish", "Base", "Method", "Type", "Flavor"]]
dricopycopy.head()

Name                                        Ingredients  \
0        '61 Imperial  1/2 oz white rum, 1/2 oz vodka, 1/2 oz 151 pro...   
1         .50 Caliber  2 oz jack daniel's® tennessee whiskey, 2 oz ji...   
2                 007  3 oz stoli® ohranj vodka, 1 1/2 ozfreshly sque...   
3             10 Deep  1.5 oz vodka, 1 oz malibu® coconut rum, 1/2 cu...   
4  100 Miles per Hour  1/2 oz wild turkey® bourbon whiskey, 1/2 oz ba...   

                                              Making  \
0  In a tall chimney glass or hurricane glass, fi...   
1  Pour the Jack Daniel's Tennessee whiskey, Jim ...   
2                  Serve on ice in a highball glass.   
3  All right start with 1/2 cup of ice mix the vo...   
4  Mix all ingredients together over ice in a hig...   

                                           Link            Glass  \
0     http://www.drinksmixer.com/drink8428.html  Hurricane Glass   
1     http://www.drinksmixer.com/drink9993.html   Highball Glass   
2    http://www.drinksmixer.com/drink11090.html   Highball Glass   
3  http://www.drinksmixer.com/drinkuv17048.html              Cup   
4    http://www.drinksmixer.com/drink10390.html   Highball Glass   

                                             Garnish                 Base  \
0   a pineapple slice and a maraschino cherry, an...   Rum, Vodka, Liquor   
1                                               None   Vodka, Whisky, Gin   
2                                               None                Vodka   
3                                               None           Rum, Vodka   
4                                               None  Rum, Whisky, Liquor   

        Method  Type Flavor  
0  Stir, Float  Long   None  
1         None  Long   None  
2         None  Long   None  
3         None         None  
4        Shake  Long   None

데이터 저장하기

In [777]:
dricopycopy.to_csv("//Drinkmixer_ready.csv", encoding="utf-8-sig", index=False)

Drink Mixer 데이터 전처리 완료